cpp/src/hash/concurrent_unordered_map.cuh

```diff
@@ -17,374 +17,423 @@
-   constexpr int block_size = 128;
-   {
-     cudaPointerAttributes hashtbl_values_ptr_attributes;
-     cudaError_t status = cudaPointerGetAttributes(
-         &hashtbl_values_ptr_attributes, m_hashtbl_values);
-
-     if (cudaSuccess == status && isPtrManaged(hashtbl_values_ptr_attributes)) {
-       int dev_id = 0;
-       CUDA_RT_CALL(cudaGetDevice(&dev_id));
-       CUDA_RT_CALL(cudaMemPrefetchAsync(
-           m_hashtbl_values, m_capacity * sizeof(value_type), dev_id, 0));
-     }
-   }
-
-   init_hashtbl<<<((m_capacity - 1) / block_size) + 1, block_size>>>(
-       m_hashtbl_values, m_capacity, m_unused_key, m_unused_element);
-   CUDA_RT_CALL(cudaGetLastError());
-   CUDA_RT_CALL(cudaStreamSynchronize(0));
-    }
-    
-    ~concurrent_unordered_map()
-    {
-        m_allocator.deallocate( m_hashtbl_values, m_capacity );
+          typename Allocator = legacy_allocator<thrust::pair<Key, Element>>>
+class concurrent_unordered_map : public managed {
+ public:
+  using size_type = size_t;
+  using hasher = Hasher;
+  using key_equal = Equality;
+  using allocator_type = Allocator;
+  using key_type = Key;
+  using mapped_type = Element;
+  using value_type = thrust::pair<Key, Element>;
+  using iterator = cycle_iterator_adapter<value_type*>;
+  using const_iterator = const cycle_iterator_adapter<value_type*>;
+
+ public:
+  /**---------------------------------------------------------------------------*
+   * @brief Construct new concurrent unordered map with of a specified
+   *m_capacity.
+   *
+   * @note The implementation of this unordered_map uses sentinel values to
+   * indicate an entry in the hash table that is empty, i.e., if a hash bucket
+   *is empty, the pair residing there will be equal to (unused_key,
+   *unused_element). As a result, attempting to insert a key equal to
+   *`unused_key` results in undefined behavior.
+   *
+   * @param _capacity The desired m_capacity of the hash table
+   * @param unused_element The sentinel value to use for an empty value
+   * @param unused_key The sentinel value to use for an empty key
+   * @param hf The hash function to use for hashing keys
+   * @param eql The equality comparison function for comparing if two keys are
+   * equal
+   * @param allocator The allocator to use for allocation the hash table's
+   * storage
+   *---------------------------------------------------------------------------**/
+  static auto create(
+      size_type capacity,
+      const mapped_type unused_element = std::numeric_limits<key_type>::max(),
+      const key_type unused_key = std::numeric_limits<key_type>::max(),
+      const Hasher& hf = hasher(), const Equality& eql = key_equal(),
+      const allocator_type& allocator = allocator_type()) {
+    using Self =
+        concurrent_unordered_map<Key, Element, Hasher, Equality, Allocator>;
+
+    auto deleter = [](Self* p) { p->destroy(); };
+
+    return std::unique_ptr<Self, decltype(deleter)>{
```

Comment:
The decltype here seems to be hitting an nvcc bug. It's causing the second argument to unique_ptr to be empty. A workaround is to use `std::function<void(Self*)>` instead.